.NET 6: error MSB4259: Unexpected space at position N Did you forget to remove a space?

### Issue Description

I'm guessing this was introduced in: https://github.com/dotnet/msbuild/pull/5672

Our build fails with:

```
error MSB4259: Unexpected space at position "47" of condition "$(RestoreTaskAssemblyFile.EndsWith('NuGet.exe', StringComparison.InvariantCultureIgnoreCase))". Did you forget to remove a space?
```

The property in question, I can fix by doing:

```diff
-<_IsRunningNuGetRestore Condition="$(RestoreTaskAssemblyFile.EndsWith('NuGet.exe', StringComparison.InvariantCultureIgnoreCase))">True</_IsRunningNuGetRestore>
+<_IsRunningNuGetRestore Condition="$(RestoreTaskAssemblyFile.EndsWith('NuGet.exe',StringComparison.InvariantCultureIgnoreCase))">True</_IsRunningNuGetRestore>
```

But then I just hit the next location:

```diff
-<AndroidSupportedHostJitAbisForConditionalChecks Condition=" !$(AndroidSupportedHostJitAbisForConditionalChecks.EndsWith (':')) "   >$(AndroidSupportedHostJitAbisForConditionalChecks):</AndroidSupportedHostJitAbisForConditionalChecks>
+<AndroidSupportedHostJitAbisForConditionalChecks Condition=" !$(AndroidSupportedHostJitAbisForConditionalChecks.EndsWith(':')) "   >$(AndroidSupportedHostJitAbisForConditionalChecks):</AndroidSupportedHostJitAbisForConditionalChecks>
```

I'm thinking these should be valid, as lots of our existing MSBuild targets would probably break.

### Steps to Reproduce

1. Use: https://dotnetcli.azureedge.net/dotnet/Sdk/6.0.100-alpha.1.20554.10/dotnet-sdk-6.0.100-alpha.1.20554.10-win-x64.exe
2. `dotnet build` this `foo.proj` file:

```xml
<Project>
  <PropertyGroup>
    <Foo>ABCD</Foo>
  </PropertyGroup>
  <ItemGroup Condition="$(Foo.StartsWith ('A'))">
    <Bar Include="True" />
  </ItemGroup>
  <Target Name="Build">
    <Message Importance="High" Text="Foo: $(Foo) Bar: @(Bar)" />
  </Target>
</Project>
```

### Expected Behavior

Build succeeds.

### Actual Behavior

Build fails with:

```
foo.proj(5,14): error MSB4259: Unexpected space at position "16" of condition "$(Foo.StartsWith ('A'))". Did you forget to remove a space?
```

### Analysis

Only guess is something in: https://github.com/dotnet/msbuild/pull/5672

### Versions & Configurations

```
> dotnet --version
6.0.100-alpha.1.20554.10
```

### Attach a binlog

[msbuild.zip](https://github.com/dotnet/msbuild/files/5497384/msbuild.zip)

--- a/src/Build.UnitTests/Scanner_Tests.cs
+++ b/src/Build.UnitTests/Scanner_Tests.cs
@@ -121,6 +121,7 @@ namespace Microsoft.Build.UnitTests
         [InlineData("$(x.StartsWith ('y'))")]
         [InlineData("$( x.StartsWith( $(SpacelessProperty) ) )")]
         [InlineData("$( x.StartsWith( $(_SpacelessProperty) ) )")]
+        [InlineData("$(x.StartsWith('Foo', StringComparison.InvariantCultureIgnoreCase))")]
         public void SpaceInMiddleOfProperty(string pattern)
         {
             Scanner lexer = new Scanner(pattern, ParserOptions.AllowProperties);

--- a/src/Build.UnitTests/Scanner_Tests.cs
+++ b/src/Build.UnitTests/Scanner_Tests.cs
@@ -6,6 +6,7 @@ using System;
 using Microsoft.Build.Evaluation;
 using Microsoft.Build.Exceptions;
 using Microsoft.Build.Utilities;
+using Shouldly;
 using Xunit;
 
 
@@ -104,16 +105,27 @@ namespace Microsoft.Build.UnitTests
         /// <summary>
         /// Tests the space errors case
         /// </summary>
-        [Fact]
-        public void SpaceProperty()
+        [Theory]
+        [InlineData("$(x )")]
+        [InlineData("$( x)")]
+        public void SpaceProperty(string pattern)
         {
-            Scanner lexer = new Scanner("$(x )", ParserOptions.AllowProperties);
+            Scanner lexer = new Scanner(pattern, ParserOptions.AllowProperties);
             AdvanceToScannerError(lexer);
             Assert.Equal("IllFormedPropertySpaceInCondition", lexer.GetErrorResource());
+        }
 
-            lexer = new Scanner("$( x)", ParserOptions.AllowProperties);
+        /// <summary>
+        /// Tests the space not next to end so no errors case
+        /// </summary>
+        [Theory]
+        [InlineData("$(x x)")]
+        [InlineData("$(y d f  ( hi ) sx)")]
+        public void SpaceInMiddleOfProperty(string pattern)
+        {
+            Scanner lexer = new Scanner(pattern, ParserOptions.AllowProperties);
             AdvanceToScannerError(lexer);
-            Assert.Equal("IllFormedPropertySpaceInCondition", lexer.GetErrorResource());
+            lexer._errorState.ShouldBeFalse();
         }
 
         [Fact]

--- a/src/Build.UnitTests/Scanner_Tests.cs
+++ b/src/Build.UnitTests/Scanner_Tests.cs
@@ -68,11 +68,7 @@ namespace Microsoft.Build.UnitTests
         /// <param name="lexer"></param>
         private void AdvanceToScannerError(Scanner lexer)
         {
-            while (true)
-            {
-                if (!lexer.Advance()) break;
-                if (lexer.IsNext(Token.TokenType.EndOfInput)) break;
-            }
+            while (lexer.Advance() && !lexer.IsNext(Token.TokenType.EndOfInput)) ;
         }
 
         /// <summary>

--- a/src/Build.UnitTests/Scanner_Tests.cs
+++ b/src/Build.UnitTests/Scanner_Tests.cs
@@ -68,7 +68,7 @@ namespace Microsoft.Build.UnitTests
         /// <param name="lexer"></param>
         private void AdvanceToScannerError(Scanner lexer)
         {
-            while (lexer.Advance() && !lexer.IsNext(Token.TokenType.EndOfInput)) ;
+            while (lexer.Advance() && !lexer.IsNext(Token.TokenType.EndOfInput));
         }
 
         /// <summary>

--- a/src/Build.UnitTests/Scanner_Tests.cs
+++ b/src/Build.UnitTests/Scanner_Tests.cs
@@ -104,6 +104,8 @@ namespace Microsoft.Build.UnitTests
         [Theory]
         [InlineData("$(x )")]
         [InlineData("$( x)")]
+        [InlineData("$([MSBuild]::DoSomething($(space ))")]
+        [InlineData("$([MSBuild]::DoSomething($(_space ))")]
         public void SpaceProperty(string pattern)
         {
             Scanner lexer = new Scanner(pattern, ParserOptions.AllowProperties);
@@ -115,8 +117,10 @@ namespace Microsoft.Build.UnitTests
         /// Tests the space not next to end so no errors case
         /// </summary>
         [Theory]
-        [InlineData("$(x x)")]
-        [InlineData("$(y d f  ( hi ) sx)")]
+        [InlineData("$(x.StartsWith( 'y' ))")]
+        [InlineData("$(x.StartsWith ('y'))")]
+        [InlineData("$( x.StartsWith( $(SpacelessProperty) ) )")]
+        [InlineData("$( x.StartsWith( $(_SpacelessProperty) ) )")]
         public void SpaceInMiddleOfProperty(string pattern)
         {
             Scanner lexer = new Scanner(pattern, ParserOptions.AllowProperties);

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -325,26 +325,22 @@ namespace Microsoft.Build.Evaluation
             {
                 fixed (char* pchar = expression)
                 {
+                    if (expression.Length > 1 && pchar[0] == '(' && char.IsWhiteSpace(pchar[1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                    {
+                        indexResult = 1;
+                        return false;
+                    }
+
                     while (index < expression.Length)
                     {
                         char character = pchar[index];
                         if (character == '(')
                         {
                             nestLevel++;
-                            if (index + 1 < expression.Length && char.IsWhiteSpace(pchar[index + 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
-                            {
-                                indexResult = index + 1;
-                                return false;
-                            }
                         }
                         else if (character == ')')
                         {
                             nestLevel--;
-                            if (index > 0 && char.IsWhiteSpace(pchar[index - 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
-                            {
-                                indexResult = index - 1;
-                                return false;
-                            }
                         }
 
                         // We have reached the end of the parenthesis nesting
@@ -352,6 +348,12 @@ namespace Microsoft.Build.Evaluation
                         // If it is not then the calling code will determine that
                         if (nestLevel == 0)
                         {
+                            if (index > 0 && char.IsWhiteSpace(pchar[index - 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                            {
+                                indexResult = index - 1;
+                                return false;
+                            }
+
                             indexResult = index;
                             return true;
                         }

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -27,7 +27,7 @@ namespace Microsoft.Build.Evaluation
         private string _expression;
         private int _parsePoint;
         private Token _lookahead;
-        private bool _errorState;
+        internal bool _errorState;
         private int _errorPosition;
         // What we found instead of what we were looking for
         private string _unexpectedlyFound = null;

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -325,7 +325,7 @@ namespace Microsoft.Build.Evaluation
             {
                 fixed (char* pchar = expression)
                 {
-                    if (expression.Length > 1 && pchar[0] == '(' && char.IsWhiteSpace(pchar[1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                    if (expression.Length > index + 1 && pchar[index] == '(' && char.IsWhiteSpace(pchar[index + 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
                     {
                         indexResult = 1;
                         return false;

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -321,8 +321,6 @@ namespace Microsoft.Build.Evaluation
         private static bool ScanForPropertyExpressionEnd(string expression, int index, out int indexResult)
         {
             int nestLevel = 0;
-            bool whitespaceCheck = false;
-
             unsafe
             {
                 fixed (char* pchar = expression)
@@ -333,17 +331,20 @@ namespace Microsoft.Build.Evaluation
                         if (character == '(')
                         {
                             nestLevel++;
-                            whitespaceCheck = true;
+                            if (index + 1 < expression.Length && char.IsWhiteSpace(pchar[index + 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                            {
+                                indexResult = index + 1;
+                                return false;
+                            }
                         }
                         else if (character == ')')
                         {
                             nestLevel--;
-                            whitespaceCheck = false;
-                        }
-                        else if (whitespaceCheck && char.IsWhiteSpace(character) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
-                        {
-                            indexResult = index;
-                            return false;
+                            if (index > 0 && char.IsWhiteSpace(pchar[index - 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                            {
+                                indexResult = index - 1;
+                                return false;
+                            }
                         }
 
                         // We have reached the end of the parenthesis nesting

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -321,16 +321,13 @@ namespace Microsoft.Build.Evaluation
         private static bool ScanForPropertyExpressionEnd(string expression, int index, out int indexResult)
         {
             int nestLevel = 0;
+            bool whitespaceFound = false;
+            bool nonIdentifierCharacterFound = false;
+            indexResult = -1;
             unsafe
             {
                 fixed (char* pchar = expression)
                 {
-                    if (expression.Length > index + 1 && pchar[index] == '(' && char.IsWhiteSpace(pchar[index + 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
-                    {
-                        indexResult = 1;
-                        return false;
-                    }
-
                     while (index < expression.Length)
                     {
                         char character = pchar[index];
@@ -342,15 +339,32 @@ namespace Microsoft.Build.Evaluation
                         {
                             nestLevel--;
                         }
+                        else if (char.IsWhiteSpace(character))
+                        {
+                            whitespaceFound = true;
+                            indexResult = index;
+                        }
+                        else if (!char.IsLetterOrDigit(character) && character != '_')
+                        {
+                            nonIdentifierCharacterFound = true;
+                        }
+
+                        if (character == '$')
+                        {
+                            if (!ScanForPropertyExpressionEnd(expression, index + 1, out index))
+                            {
+                                indexResult = index;
+                                return false;
+                            }
+                        }
 
                         // We have reached the end of the parenthesis nesting
                         // this should be the end of the property expression
                         // If it is not then the calling code will determine that
                         if (nestLevel == 0)
                         {
-                            if (index > 0 && char.IsWhiteSpace(pchar[index - 1]) && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
+                            if (whitespaceFound && !nonIdentifierCharacterFound && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave16_10))
                             {
-                                indexResult = index - 1;
                                 return false;
                             }
 

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -344,7 +344,7 @@ namespace Microsoft.Build.Evaluation
                             whitespaceFound = true;
                             indexResult = index;
                         }
-                        else if (!char.IsLetterOrDigit(character) && character != '_')
+                        else if (!XmlUtilities.IsValidSubsequentElementNameCharacter(character))
                         {
                             nonIdentifierCharacterFound = true;
                         }

--- a/src/Build/Evaluation/Conditionals/Scanner.cs
+++ b/src/Build/Evaluation/Conditionals/Scanner.cs
@@ -349,7 +349,7 @@ namespace Microsoft.Build.Evaluation
                             nonIdentifierCharacterFound = true;
                         }
 
-                        if (character == '$')
+                        if (character == '$' && index < expression.Length - 1 && pchar[index + 1] == '(')
                         {
                             if (!ScanForPropertyExpressionEnd(expression, index + 1, out index))
                             {

